File: src/services/spotifyService.js
import SpotifyWebApi from 'spotify-web-api-node';
import { config } from '../config.js';
import { Logger } from '../utils/logger.js';

class SpotifyService {
  constructor() {
    if (!config.spotify.clientId || !config.spotify.clientSecret) {
      Logger.warn('Spotify credentials not configured. Spotify support disabled.');
      this.enabled = false;
      return;
    }

    // Validate credentials format
    if (config.spotify.clientId.trim() === '' || config.spotify.clientSecret.trim() === '') {
      Logger.warn('Spotify credentials are empty. Spotify support disabled.');
      this.enabled = false;
      return;
    }

    this.spotify = new SpotifyWebApi({
      clientId: config.spotify.clientId,
      clientSecret: config.spotify.clientSecret,
    });
    
    this.enabled = true;
    this.authenticate();
  }

  async authenticate() {
    if (!this.enabled) {
      return;
    }

    try {
      const data = await this.spotify.clientCredentialsGrant();
      this.spotify.setAccessToken(data.body.access_token);
      Logger.info('Spotify API authenticated successfully');
      
      // Refresh token before it expires
      setTimeout(() => this.authenticate(), (data.body.expires_in - 300) * 1000);
    } catch (error) {
      Logger.warn('Spotify authentication failed - check your credentials in .env file');
      Logger.warn('Spotify features will be disabled. YouTube search will still work.');
      this.enabled = false;
    }
  }

  isSpotifyUrl(url) {
    return url.includes('spotify.com/playlist/') || url.includes('open.spotify.com/playlist/');
  }

  extractPlaylistId(url) {
    const match = url.match(/playlist\/([a-zA-Z0-9]+)/);
    return match ? match[1] : null;
  }

  async getPlaylistTracks(playlistId) {
    if (!this.enabled) {
      throw new Error('Spotify service is not enabled');
    }

    try {
      const playlist = await this.spotify.getPlaylist(playlistId);
      const tracks = [];
      
      let offset = 0;
      const limit = 100;
      let hasMore = true;

      while (hasMore) {
        const response = await this.spotify.getPlaylistTracks(playlistId, {
          offset,
          limit,
          fields: 'items(track(name,artists(name),duration_ms)),next'
        });

        const items = response.body.items.filter(item => item.track && item.track.name);
        
        tracks.push(...items.map(item => ({
          title: item.track.name,
          artist: item.track.artists.map(artist => artist.name).join(', '),
          duration: item.track.duration_ms,
          searchQuery: `${item.track.name} ${item.track.artists[0].name}`
        })));

        hasMore = response.body.next !== null;
        offset += limit;
      }

      Logger.info(`Found ${tracks.length} tracks in Spotify playlist: ${playlist.body.name}`);
      return {
        name: playlist.body.name,
        tracks
      };
    } catch (error) {
      Logger.error('Failed to fetch Spotify playlist:', error.message);
      throw new Error('Failed to fetch Spotify playlist');
    }
  }
}

export const spotifyService = new SpotifyService();

File: src/services/spotify_service.py
"""
Spotify service for handling Spotify playlist integration
"""

import spotipy
from spotipy.oauth2 import SpotifyClientCredentials
import logging
from src.config import config
from src.utils.logger import get_logger

logger = get_logger(__name__)

class SpotifyService:
    """Spotify service for playlist integration"""
    
    def __init__(self):
        if not config.SPOTIFY_CLIENT_ID or not config.SPOTIFY_CLIENT_SECRET:
            logger.warning("Spotify credentials not configured. Spotify support disabled.")
            self.enabled = False
            return
        
        # Validate credentials format
        if not config.SPOTIFY_CLIENT_ID.strip() or not config.SPOTIFY_CLIENT_SECRET.strip():
            logger.warning("Spotify credentials are empty. Spotify support disabled.")
            self.enabled = False
            return
        
        try:
            # Initialize Spotify client
            client_credentials_manager = SpotifyClientCredentials(
                client_id=config.SPOTIFY_CLIENT_ID,
                client_secret=config.SPOTIFY_CLIENT_SECRET
            )
            
            self.spotify = spotipy.Spotify(
                client_credentials_manager=client_credentials_manager
            )
            
            self.enabled = True
            logger.info("Spotify API authenticated successfully")
            
        except Exception as error:
            logger.warning(f"Spotify authentication failed: {error}")
            logger.warning("Spotify features will be disabled. YouTube search will still work.")
            self.enabled = False
    
    def is_spotify_url(self, url):
        """Check if URL is a Spotify playlist URL"""
        return 'spotify.com/playlist/' in url or 'open.spotify.com/playlist/' in url
    
    def extract_playlist_id(self, url):
        """Extract playlist ID from Spotify URL"""
        import re
        match = re.search(r'playlist/([a-zA-Z0-9]+)', url)
        return match.group(1) if match else None
    
    async def get_playlist_tracks(self, playlist_id):
        """Get tracks from a Spotify playlist"""
        if not self.enabled:
            raise Exception("Spotify service is not enabled")
        
        try:
            # Get playlist info
            playlist = self.spotify.playlist(playlist_id)
            tracks = []
            
            # Get all tracks (handle pagination)
            results = self.spotify.playlist_tracks(playlist_id)
            
            while results:
                for item in results['items']:
                    if item['track'] and item['track']['name']:
                        track = item['track']
                        artists = ', '.join([artist['name'] for artist in track['artists']])
                        
                        tracks.append({
                            'title': track['name'],
                            'artist': artists,
                            'duration': track['duration_ms'],
                            'search_query': f"{track['name']} {track['artists'][0]['name']}"
                        })
                
                # Get next page if available
                results = self.spotify.next(results) if results['next'] else None
            
            logger.info(f"Found {len(tracks)} tracks in Spotify playlist: {playlist['name']}")
            
            return {
                'name': playlist['name'],
                'tracks': tracks
            }
            
        except Exception as error:
            logger.error(f"Failed to fetch Spotify playlist: {error}")
            raise Exception("Failed to fetch Spotify playlist")